mypy/checker.py

```diff
@@ -1142,33 +1134,49 @@ def check_rvalue_count_in_assignment(self, lvalues: List[Node], rvalue_count: in
 
-    def check_multi_assignment(self, lvalues: List[Node],
-                               rvalue: Node,
-                               context: Context,
-                               infer_lvalue_type: bool = True,
-                               msg: str = None) -> None:
-        """Check the assignment of one rvalue to a number of lvalues."""
-
-        # Infer the type of an ordinary rvalue expression.
-        rvalue_type = self.accept(rvalue)  # TODO maybe elsewhere; redundant
-        undefined_rvalue = False
+    def check_multi_assign(self, lvalues: List[Expression],
+                           rvalue: Expression, rvalue_type: Type,
+                           context: Context, infer_lvalue_type: bool = True,
+                           undefined_rvalue = False) -> None:
 
         if isinstance(rvalue_type, AnyType):
             for lv in lvalues:
                 if isinstance(lv, StarExpr):
                     lv = lv.expr
                 self.check_assignment(lv, self.temp_node(AnyType(), context), infer_lvalue_type)
         elif isinstance(rvalue_type, TupleType):
-            self.check_multi_assignment_from_tuple(lvalues, rvalue, rvalue_type,
-                                                  context, undefined_rvalue, infer_lvalue_type)
+            self.check_multi_assign_from_tuple(lvalues, rvalue, rvalue_type,
+                                               context, undefined_rvalue, infer_lvalue_type)
+        elif isinstance(rvalue_type, UnionType):
+            self.check_multi_assign_from_union(lvalues, rvalue, rvalue_type,
+                                               context, infer_lvalue_type)
+        elif isinstance(rvalue_type, Instance) and self.type_is_iterable(rvalue_type):
+            self.check_multi_assign_from_iterable(lvalues, rvalue_type,
+                                                  context, infer_lvalue_type)
         else:
-            self.check_multi_assignment_from_iterable(lvalues, rvalue_type,
-                                                     context, infer_lvalue_type)
+            self.msg.type_not_iterable(rvalue_type, context)
+
+    def check_multi_assign_from_union(self, lvalues: List[Expression], rvalue: Expression,
+                                      rvalue_type: UnionType, context: Context,
+                                      infer_lvalue_type: bool) -> None:
+        union_types = tuple([] for _ in lvalues)  # type: Tuple[List[Type], ...]
+        for item in rvalue_type.items:
+            self.check_multi_assign(lvalues, rvalue, item, context, infer_lvalue_type,
+                                    undefined_rvalue=True)
+            for t, lv in zip(union_types, lvalues):
+                t.append(self.type_map[lv])
+        for ut, lv in zip(union_types, lvalues):
+            _1, _2, inferred = self.check_lvalue(lv)
+            union = join_type_list(ut)
+            if inferred:
+                self.set_inferred_type(inferred, lv, union)
```

Comment:
You'll probably need to put the new type to the binder (maybe `self.binder.assign_type(...)`).  Also add a test case for this. Example where this may make a difference:

```
class A: pass
class B(A): pass
class C(A): pass
x = None # type: object
a = [] # type: Union[List[B], List[C]]
x, y = a
# now type of x should be A
x = 1  # should be ok
# now type of x should be int
```